agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+        let mut tasks = Vec::new();
+        let discovery_operator = Arc::new(discovery_operator);
+        let task1_discovery_operator = discovery_operator.clone();
+        // Call discover on already registered Discovery Handlers for this Configuration's protocol
+        tasks.push(tokio::spawn(async move {
+            do_discover(
+                task1_discovery_operator,
+                Arc::new(Box::new(k8s::create_kube_interface())),
+            )
+            .await
+            .unwrap();
+        }));
+        let mut stop_all_discovery_receiver = stop_all_discovery_sender.subscribe();
+        let mut new_discovery_handler_receiver = new_discovery_handler_sender.subscribe();
+        let task2_discovery_operator = discovery_operator.clone();
+        tasks.push(tokio::spawn(async move {
+            let mut inner_tasks = Vec::new();
+            loop {
+                tokio::select! {
+                    _ = try_receive(&mut stop_all_discovery_receiver) => {
+                        trace!("start_discovery - received message to stop discovery for configuration {}", task2_discovery_operator.get_config().metadata.name);
+                        // stop_offline_checks_sender.send(()).unwrap();
+                        task2_discovery_operator.stop_all_discovery().await;
+                        break;
+                    },
+                    result = try_receive(&mut new_discovery_handler_receiver) => {
+                        // check if it is this protocol
+                        if let Ok(protocol) = result {
+                            if protocol == task2_discovery_operator.get_config().spec.protocol.name {
+                                trace!("start_discovery - received new registered discovery handler for configuration {}", task2_discovery_operator.get_config().metadata.name);
+                                let new_discovery_operator = task2_discovery_operator.clone();
+                                inner_tasks.push(tokio::spawn(async move {
+                                    do_discover(new_discovery_operator, Arc::new(Box::new(k8s::create_kube_interface()))).await.unwrap();
+                                }));
+                            }
+                        }
+                    }
+                }
+            }
+            futures::future::try_join_all(inner_tasks).await.unwrap();
+        }));
+        let mut stop_all_discovery_receiver = stop_all_discovery_sender.subscribe();
+        let task3_discovery_operator = discovery_operator.clone();
+        // Non-local devices are only allowed to be offline for `SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS` minutes before being removed.
+        // This task periodically checks if devices have been offline for too long.
+        tasks.push(tokio::spawn(async move {
+            let kube_interface: Arc<Box<dyn k8s::KubeInterface>> = Arc::new(Box::new(k8s::create_kube_interface()));
+            loop {
```

Comment:
Is it cheaper to substitute that with a timer that is only started with (SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS) when a device is marked as offline (line 458)?

Reply:
I was wondering about this too but I worry about this scenario, assuming a sleep time of 60 seconds between checks in spawned status check thread: device X goes offline, thread X spawned for it, 10 seconds later device X reported back online, 10 seconds later device X goes offline again, thread X' spawned for it. If the device continues to go online and offline in these short intervals, a scenario could happen where more and more threads are being spawned and never ended if the threads are only waking when the `InstanceConnectivityStatus` is reset to Offline.

Reply:
I see your point. It can vary depending on the situation.